### DISCUSSION TITLE
Unknown RAN Function ID

### DISCUSSION THREAD

**Discussion by LVancina** [srsran/srsRAN_Project#368] [General Help]
Greetings,

I am new to srsRAN and I am trying to get one of the example experiments from the docs up and running on my Dell Precision 7670 running Ubuntu 22.04. I followed all of the instructions found here: https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html, and I can start up the Open5gc core and FlexRIC without a problem, with the following output initially on the FlexRIC terminal

lucas@lucas-dell-research:~$ ./flexric/build/examples/ric/nearRT-RIC
Setting the config -c file to /usr/local/etc/flexric/flexric.conf
Setting path -p for the shared libraries to /usr/local/lib/flexric/
[NEAR-RIC]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36421
[NEAR-RIC]: Initializing 
[NEAR-RIC]: Loading SM ID = 142 with def = MAC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 148 with def = GTP_STATS_V0 
[NEAR-RIC]: Loading SM ID = 145 with def = SLICE_STATS_V0 
[NEAR-RIC]: Loading SM ID = 144 with def = PDCP_STATS_V0 
[NEAR-RIC]: Loading SM ID = 143 with def = RLC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 147 with def = ORAN-E2SM-KPM 
[NEAR-RIC]: Loading SM ID = 146 with def = TC_STATS_V0 
[iApp]: Initializing ... 
[iApp]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36422
fd created with 6

but, when I start up the gNB, I get the following output in the FlexRIC terminal:

Received message with id = 411, port = 40429 
[E2AP] Received SETUP-REQUEST from PLMN   1. 1 Node ID 411 RAN type ngran_gNB
Unknown RAN function ID, thus rejecting 2 
nearRT-RIC: /home/lucas/flexric/src/ric/msg_handler_ric.c:100: generate_setup_response: Assertion `0!=0 && "Unknown RAN function ID from the agent received"' failed.
Aborted (core dumped)
lucas@lucas-dell-research:~$ 

For reference, this is what the ouput in the gNB terminal looks like:

lucas@lucas-dell-research:~/srsRAN_Project$ sudo ./build/apps/gnb/gnb -c gnb_zmq.yaml 

The PRACH detector will not meet the performance requirements with the configuration {Format 0, ZCZ 0, SCS 1.25kHz, Rx ports 1}.
Lower PHY in executor blocking mode.

--== srsRAN gNB (commit 55c984b55) ==--

Connecting to AMF on 10.53.1.2:38412
Available radio types: zmq.
Connecting to NearRT-RIC on 127.0.0.1:36421
Cell pci=1, bw=10 MHz, dl_arfcn=368500 (n3), dl_freq=1842.5 MHz, dl_ssb_arfcn=367930, ul_freq=1747.5 MHz

==== gNodeB started ===
Type <t> to view trace

Can anyone give me a hint about why I might be having this issue? I had this same experiment running just fine on a VM, but it's giving me problems now that I have a dedicated Linux machine.

Please let me know what further details might be helpful.

**Comment by pgawlowicz** [comment]
Hi, thanks for pointing this out. After the last code release, we need to release also the new version of the app note.

**Reply by LVancina** [reply]
It sounds like this is a documentation issue with the new code release. Is there a known fix-action?

**Comment by pengnist** [comment]
We also had this issue with the latest updates. Also, when we tried running the testbed without flexric with the latest open5gs and srsran_project gnb and srsran_4g ue, connection could not be established. Is this a known issue?

**Comment by LukasWehrstein** [comment]
The problem is still not solved, however I found out, that when using the latest flexric version , at least the RIC does not crash when starting the srsRAN gNB.

Follow the installation guide for flexric (I also needed to update swig to v4.1)
https://gitlab.eurecom.fr/mosaic5g/flexric#1-installation

And then build the project using these flags in the CMAKE command
```bash
CC=gcc-10 CXX=g++-10 cmake .. -DE2AP_VERSION=E2AP_V3 -DKPM_VERSION=KPM_V3_00
make install
```

After starting the open5GS core from the srsRAN docker and the nearRT-RIC I get this output (not crashing) in the RIC:
```
[UTIL]: Setting the config -c file to /usr/local/etc/flexric/flexric.conf
[UTIL]: Setting path -p for the shared libraries to /usr/local/lib/flexric/
[NEAR-RIC]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36421
[NEAR-RIC]: Initializing
[NEAR-RIC]: Loading SM ID = 143 with def = RLC_STATS_V0
[NEAR-RIC]: Loading SM ID = 142 with def = MAC_STATS_V0
[NEAR-RIC]: Loading SM ID = 145 with def = SLICE_STATS_V0
[NEAR-RIC]: Loading SM ID = 3 with def = ORAN-E2SM-RC-v01.0
[NEAR-RIC]: Loading SM ID = 146 with def = TC_STATS_V0
[NEAR-RIC]: Loading SM ID = 144 with def = PDCP_STATS_V0
[NEAR-RIC]: Loading SM ID = 2 with def = ORAN-E2SM-KPM-v03.00
[NEAR-RIC]: Loading SM ID = 148 with def = GTP_STATS_V0
[iApp]: Initializing ...
[iApp]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36422
[NEAR-RIC]: Initializing Task Manager with 2 threads
[E2AP]: E2 SETUP-REQUEST rx from PLMN   1. 1 Node ID 411 RAN type ngran_gNB
[NEAR-RIC]: Accepting RAN function ID 2 with def = ORAN-E2SM-KPM-v03.00
```

This means, now the Functions with SM ID = 2 and 3 are known to the RIC and the srsRAN gNB is able to connect via E2 Interface.

Here the output of the gNB:
```bash
./gnb -c gnb_zmq.yaml

The PRACH detector will not meet the performance requirements with the configuration {Format 0, ZCZ 0, SCS 1.25kHz, Rx ports 1}.
Lower PHY in executor blocking mode.

--== srsRAN gNB (commit 55c984b55) ==--

Connecting to AMF on 10.53.1.2:38412
Available radio types: zmq.
Connecting to NearRT-RIC on 127.0.0.1:36421
Cell pci=1, bw=10 MHz, dl_arfcn=368500 (n3), dl_freq=1842.5 MHz, dl_ssb_arfcn=367930, ul_freq=1747.5 MHz

==== gNodeB started ===
Type <t> to view trace
```

However, when i start the UE now, i do not get a connection to the gnb via zmq - but this might be related to another issue - i dont know.

Maybe this helps some of you guys - looking forward to hear from you, if anyone has the same issue or a solution for getting the UE connected.

### Update 08.01.2024 - ZMQ working

After copying the adjustments on the config files for gNB and UE from thread https://github.com/srsran/srsRAN_Project/discussions/350, I am now able to use the setup with zmq.

**Reply by pengnist** [reply]
Was the connection established when using USRPs? If so, do you mind sharing the gains on both gnb and ue, and the attenuation in between?
I had issues using docker compose to run open5gs on Ubuntu 22.04, and the latest Open5GS installed from repo does not work. Not sure if it is the open5gs or the gain configs.

**Reply by pengnist** [reply]
@LVancina I tried adding the flags to the cmake and can verify that it works! I used the docker open5gs that srsran project provided (the open5gs version in it is v2.6.1) and used USRPs. Connection is established between gnb and srsue. 

I can also duplicate the failed connection issue using zmq. You are right that it may be related to something else.

**Reply by LVancina** [reply]
@pengnist It didn't work for me. I'm wondering if I am making a mistake by switching to the e2ap-v2 branch after cloning the flexric repo. Did you do that? If not, do you mind sharing which release tag you used and the values you used for the -DE2AP_VERSION=<e2ap_version> -DKPM_VERSION=<kpm_version> flags?

**Reply by pengnist** [reply]
I used the tag that @LukasWehrstein provided for cmake: CC=gcc-10 CXX=g++-10 cmake .. -DE2AP_VERSION=E2AP_V3 -DKPM_VERSION=KPM_V3_00. I didn't switch to e2ap-v2 branch. Instead, I used their master.
Also, I used the dockerized open5gs from srsRAN Project. I believe the version of the dockerized open5gs is v2.6.1. My guess is srs has issues with the latest open5gs.

**Comment by WirelessRe** [comment]
I am facing the same issue, and using -DE2AP_VERSION=<e2ap_version> -DKPM_VERSION=<kpm_version>  with cmake does not work

**Reply by LukasWehrstein** [reply]
Maybe double check if you are on flexric master branch.
Or - maybe you need to specify teh compiler version - as shown in my previous post - i was not able to compile it with the latest gcc...

**Comment by pengnist** [comment]
@LukasWehrstein Did you have all tests passed when running ctest? I had a few tests failed, but near-RT RIC was able to connect to gnb.

**Comment by hasanyeganeh** [comment]
![Capture](https://github.com/srsran/srsRAN_Project/assets/33173939/6a6d2aff-b802-498d-a006-85fb1a82283b)
I can see all tests have been passed but still I have "Unknown RAN function ID from the agent received" issue

**Comment by pgawlowicz** [comment]
Please check our updated [tutorial](https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html), where the above issue was solved by using a newer version of Flexric.
Specifically, we use the `br-flexric` branch that fixes the issue. Moreover, it allows specifying metric names using a config file that is passed to an xApp.

**Reply by mas-oliveira** [reply]
Hello @pgawlowicz,

In this part of the upload:

"Enable Enable RLC metrics reporting that will feed E2SM_KPM service model with measurements data:

metrics:
  rlc_json_enable: 1                # Enable RLC metrics reporting
  rlc_report_period: 1000           # Set reporting period to 1s
"

Do we need to add anything to the directory? Because when I run it, like adding that part to the gnb_zmq.conf, it gives this error:
INI was not able to parse metrics.rlc_json_enable
Run with --help for more information.

Then I run everything like the tutorial, but on the xApp, the metrics are 0 and the values of the timestamp are negatives, so I think there is an error on the part that is filling the E2 KPM message. Do you think it is better to open an issue, or is there something that is missing in the new tutorial?

Thanks

**Reply by pgawlowicz** [reply]
You need to use the newest release of the srsgnb (commit: 0b2702c). Then you need to start some traffic, e.g., with iperf:
```
sudo ip netns exec ue1 iperf -c 10.45.1.1 -u -b 10M -i 1 -t 60
```
so there is some traffic to measure the DL and UL throughput > 0
